runtime/signals.c

```diff
@@ -104,20 +104,21 @@ CAMLexport value caml_process_pending_signals_exn(void)
 
 /* Record the delivery of a signal, and arrange for it to be processed
    as soon as possible:
-   - via the pending signal bitvector, processed in
-     caml_process_pending_signals_exn.
+   - via Caml_state->action_pending, processed in
+     caml_process_pending_actions.
    - by playing with the allocation limit, processed in
-     caml_garbage_collection
+     caml_alloc_small_dispatch.
 */
-
 CAMLexport void caml_record_signal(int signal_number)
 {
   unsigned int i;
   if (signal_number <= 0 || signal_number >= NSIG) return;
   i = signal_number - 1;
   atomic_fetch_or(&caml_pending_signals[i / BITS_PER_WORD],
                   (uintnat)1 << (i % BITS_PER_WORD));
-  caml_interrupt_self();
+  // FIXME: Is it possible that Caml_state is not accessible from the
+  // current thread?
+  caml_set_action_pending();
```

Comment:
This is a correctness concern for the code currently in trunk (see 15th commit "do not access TLS in signal handler"). Let me know if it is justified.